File: C#/WindowsPhone/LGame-XNA-lib/Loon.Action.Avg/AVGChara.cs
namespace Loon.Action.Avg
{
    using System.IO;
    using Loon.Core;
    using Loon.Core.Graphics.Opengl;
    using Loon.Action.Sprite.Effect;
    using Loon.Action.Sprite;

    public class AVGChara : LRelease
    {

        private LTexture characterCG;

        private int width;

        private int height;

        internal int x;

        internal int y;

        internal int flag = -1;

        internal float time;

        internal float currentFrame;

        internal float opacity;

        protected internal bool isMove, isAnimation, isVisible = true;

        internal int maxWidth, maxHeight;

        private int moveX;

        private int direction;

        private int moveSleep = 10;

        private bool moving;

        protected internal AVGAnm anm;

        /// <summary>
        /// 构造函数，初始化角色图
        /// </summary>
        ///
        /// <param name="image"></param>
        /// <param name="x"></param>
        /// <param name="y"></param>
        /// <param name="width"></param>
        /// <param name="height"></param>
        public AVGChara(LTexture image, int x, int y, int width,
                int height)
        {
            this.Load(image, x, y, width, height, LSystem.screenRect.width,
                    LSystem.screenRect.height);
        }

        public AVGChara(LTexture image, int x, int y)
        {
            this.Load(image, x, y);
        }

        public AVGChara(string resName, int x, int y):this(resName, x, y, LSystem.screenRect.width, LSystem.screenRect.height)
        {
            
        }

        public AVGChara(string resName, int x, int y,
                int w, int h)
        {
            string path = resName;
            if (path.StartsWith("\""))
            {
                path = resName.Replace("\"", "");
            }
            if (path.EndsWith(".an"))
            {
                this.x = x;
                this.y = y;
                this.isAnimation = true;
                try
                {
                    this.anm = new AVGAnm(path);
                }
                catch (IOException e)
                {
                    Loon.Utils.Debug.Log.Exception(e.StackTrace);
                }
                this.maxWidth = w;
                this.maxHeight = h;
            }
            else
            {
                this.Load(LTextures.LoadTexture(path), x, y);
            }
        }

        internal string tmp_path;

        internal void Update(string path)
        {
            this.tmp_path = path;
        }

        private void Load(LTexture image, int x, int y)
        {
            this.Load(image, x, y, image.GetWidth(), image.GetHeight(),
                    LSystem.screenRect.width, LSystem.screenRect.height);
        }

        private void Load(LTexture image, int x, int y, int width,
                int height, int w, int h)
        {
            this.maxWidth = w;
            this.maxHeight = h;
            this.isAnimation = false;
            this.characterCG = image;
            this.isMove = true;
            this.width = width;
            this.height = height;
            this.x = x;
            this.y = y;
            this.moveX = 0;
            this.direction = GetDirection();
            if (direction == 0)
            {
                this.moveX = -(width / 2);
            }
            else
            {
                this.moveX = maxWidth;
            }
        }

        public void SetFlag(int f, float delay)
        {
            this.flag = f;
            this.time = delay;
            if (flag == ISprite_Constants.TYPE_FADE_IN)
            {
                this.currentFrame = this.time;
            }
            else
            {
                this.currentFrame = 0;
            }
        }

        public int GetScreenWidth()
        {
            return maxWidth;
        }

        public int GetScreenHeight()
        {
            return maxHeight;
        }

        private int GetDirection()
        {
            int offsetX = maxWidth / 2;
            if (x < offsetX)
            {
                return 0;
            }
            else
            {
                return 1;
            }
        }

        public void SetMove(bool move)
        {
            isMove = move;
        }

        public void Flush()
        {
            characterCG = null;
            x = 0;
            y = 0;
        }

        public int GetNext()
        {
            return moveX;
        }

        public int GetMaxNext()
        {
            return x;
        }

        public bool Next()
        {
            moving = false;
            if (moveX != x)
            {
                for (int sleep = 0; sleep < moveSleep; sleep++)
                {
                    if (direction == 0)
                    {
                        moving = (x > moveX);
                    }
                    else
                    {
                        moving = (x < moveX);
                    }
                    if (moving)
                    {
                        switch (direction)
                        {
                            case 0:
                                moveX += 1;
                                break;
                            case 1:
                                moveX -= 1;
                                break;
                            default:
                                moveX = x;
                                break;
                        }
                    }
                    else
                    {
                        moveX = x;
                    }
                }
            }
            return moving;
        }

        internal void Update(long t)
        {

        }

        internal void draw(GLEx g)
        {
            g.DrawTexture(characterCG, moveX, y);
        }

        public int GetX()
        {
            return x;
        }

        public void SetX(int x)
        {
            if (isMove)
            {
                int move = x - this.moveX;
                if (move < 0)
                {
                    this.moveX = this.x;
                    this.x = x;
                    direction = 1;
                }
                else
                {
                    this.moveX = move;
                    this.x = x;
                }
            }
            else
            {
                this.moveX = x;
                this.x = x;
            }

        }

        public int GetY()
        {
            return y;
        }

        public void SetY(int y)
        {
            this.y = y;
        }

        public int GetHeight()
        {
            return height;
        }

        public void SetHeight(int height)
        {
            this.height = height;
        }

        public int GetWidth()
        {
            return width;
        }

        public void SetWidth(int width)
        {
            this.width = width;
        }

        public int GetMoveSleep()
        {
            return moveSleep;
        }

        public void SetMoveSleep(int moveSleep)
        {
            this.moveSleep = moveSleep;
        }

        public int GetMoveX()
        {
            return moveX;
        }

        public bool IsAnimation()
        {
            return isAnimation;
        }

        internal void SetAnimation(bool isAnimation)
        {
            this.isAnimation = isAnimation;
        }

        public bool IsVisible()
        {
            return isVisible;
        }

        public void SetVisible(bool isVisible)
        {
            this.isVisible = isVisible;
        }

        public LTexture GetTexture()
        {
            return characterCG;
        }

        public void Dispose()
        {
            this.isVisible = false;
            if (characterCG != null)
            {
                characterCG.Destroy();
                characterCG = null;
            }
            if (anm != null)
            {
                anm.Dispose();
                anm = null;
                isAnimation = false;
            }
        }

    }
}

File: Java/Android/OpenGLES/source/core/src/loon/action/avg/AVGChara.java
package loon.action.avg;

import java.io.IOException;

import loon.action.sprite.ISprite;
import loon.core.LRelease;
import loon.core.LSystem;
import loon.core.graphics.opengl.GLEx;
import loon.core.graphics.opengl.LTexture;
import loon.core.graphics.opengl.LTextures;
import loon.utils.StringUtils;

/**
 * Copyright 2008 - 2010
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.1
 */
public class AVGChara implements LRelease {

	private LTexture characterCG;

	private int width;

	private int height;

	int x;

	int y;

	int flag = -1;

	float time;

	float currentFrame;

	float opacity;

	protected boolean isMove, isAnimation, isVisible = true;

	int maxWidth, maxHeight;

	private int moveX;

	private int direction;

	private int moveSleep = 10;

	private boolean moving;

	protected AVGAnm anm;

	/**
	 * 构造函数，初始化角色图
	 * 
	 * @param image
	 * @param x
	 * @param y
	 * @param width
	 * @param height
	 */
	public AVGChara(LTexture image, final int x, final int y, int width,
			int height) {
		this.load(image, x, y, width, height, LSystem.screenRect.width,
				LSystem.screenRect.height);
	}

	public AVGChara(LTexture image, final int x, final int y) {
		this.load(image, x, y);
	}

	public AVGChara(final String resName, final int x, final int y) {
		this(resName, x, y, LSystem.screenRect.width, LSystem.screenRect.height);
	}

	public AVGChara(final String resName, final int x, final int y,
			final int w, final int h) {
		String path = resName;
		if (StringUtils.startsWith(path, '"')) {
			path = resName.replaceAll("\"", "");
		}
		if (path.endsWith(".an")) {
			this.x = x;
			this.y = y;
			this.isAnimation = true;
			try {
				this.anm = new AVGAnm(path);
			} catch (IOException e) {
				e.printStackTrace();
			}
			this.maxWidth = w;
			this.maxHeight = h;
		} else {
			this.load(LTextures.loadTexture(path), x, y);
		}
	}

	String tmp_path;

	void update(String path) {
		this.tmp_path = path;
	}

	private void load(LTexture image, final int x, final int y) {
		this.load(image, x, y, image.getWidth(), image.getHeight(),
				LSystem.screenRect.width, LSystem.screenRect.height);
	}

	private void load(LTexture image, final int x, final int y, int width,
			int height, final int w, final int h) {
		this.maxWidth = w;
		this.maxHeight = h;
		this.isAnimation = false;
		this.characterCG = image;
		this.isMove = true;
		this.width = width;
		this.height = height;
		this.x = x;
		this.y = y;
		this.moveX = 0;
		this.direction = getDirection();
		if (direction == 0) {
			this.moveX = -(width / 2);
		} else {
			this.moveX = maxWidth;
		}
	}

	public void setFlag(int f, float delay) {
		this.flag = f;
		this.time = delay;
		if (flag == ISprite.TYPE_FADE_IN) {
			this.currentFrame = this.time;
		} else {
			this.currentFrame = 0;
		}
	}

	@Override
	public void finalize() {
		flush();
	}

	public int getScreenWidth() {
		return maxWidth;
	}

	public int getScreenHeight() {
		return maxHeight;
	}

	private int getDirection() {
		int offsetX = maxWidth / 2;
		if (x < offsetX) {
			return 0;
		} else {
			return 1;
		}
	}

	public void setMove(boolean move) {
		isMove = move;
	}

	public void flush() {
		characterCG = null;
		x = 0;
		y = 0;
	}

	public int getNext() {
		return moveX;
	}

	public int getMaxNext() {
		return x;
	}

	public boolean next() {
		moving = false;
		if (moveX != x) {
			for (int sleep = 0; sleep < moveSleep; sleep++) {
				if (direction == 0) {
					moving = (x > moveX);
				} else {
					moving = (x < moveX);
				}
				if (moving) {
					switch (direction) {
					case 0:
						moveX += 1;
						break;
					case 1:
						moveX -= 1;
						break;
					default:
						moveX = x;
						break;
					}
				} else {
					moveX = x;
				}
			}
		}
		return moving;
	}

	void update(long t) {

	}

	void draw(GLEx g) {
		g.drawTexture(characterCG, moveX, y);
	}

	public int getX() {
		return x;
	}

	public void setX(int x) {
		if (isMove) {
			int move = x - this.moveX;
			if (move < 0) {
				this.moveX = this.x;
				this.x = x;
				direction = 1;
			} else {
				this.moveX = move;
				this.x = x;
			}
		} else {
			this.moveX = x;
			this.x = x;
		}

	}

	public int getY() {
		return y;
	}

	public void setY(int y) {
		this.y = y;
	}

	public int getHeight() {
		return height;
	}

	public void setHeight(int height) {
		this.height = height;
	}

	public int getWidth() {
		return width;
	}

	public void setWidth(int width) {
		this.width = width;
	}

	public int getMoveSleep() {
		return moveSleep;
	}

	public void setMoveSleep(int moveSleep) {
		this.moveSleep = moveSleep;
	}

	public int getMoveX() {
		return moveX;
	}

	public boolean isAnimation() {
		return isAnimation;
	}

	void setAnimation(boolean isAnimation) {
		this.isAnimation = isAnimation;
	}

	public boolean isVisible() {
		return isVisible;
	}

	public void setVisible(boolean isVisible) {
		this.isVisible = isVisible;
	}

	public LTexture getTexture() {
		return characterCG;
	}

	@Override
	public void dispose() {
		this.isVisible = false;
		if (characterCG != null) {
			characterCG.destroy();
			characterCG = null;
		}
		if (anm != null) {
			anm.dispose();
			anm = null;
			isAnimation = false;
		}
	}

}
